File: src/payout-service/create-payout.ts
import { nanoid } from "nanoid";
import { ExpectedError, Reason } from "../errors/expected-error";
import {
	CreatePayoutRequest,
	CreatePayoutResponse,
	Payout,
	Payout_Status,
} from "../gen/ride/payout/v1alpha1/payout_service";
import PayoutRepository from "../repositories/payout-repository";
import WalletRepository from "../repositories/wallet-repository";
import razorpay from "../third-party/razorpay";
import { moneyToInt, walletRegex } from "../utils";

async function createPayout(
	request: CreatePayoutRequest
): Promise<CreatePayoutResponse> {
	try {
		if (request.parent.match(walletRegex) === null) {
			throw new ExpectedError("Invalid parent", Reason.INVALID_ARGUMENT);
		}

		if (!request.payout) {
			throw new ExpectedError("Payout is empty", Reason.INVALID_ARGUMENT);
		}

		if (
			!request.payout.amount ||
			request.payout.amount.currencyCode !== "INR" ||
			request.payout.amount.units <= 0 ||
			request.payout.amount.nanos < 0
		) {
			throw new ExpectedError("Invalid amount", Reason.INVALID_ARGUMENT);
		}

		const walletId = request.parent.split("/").pop();

		if (!walletId || walletId === "-") {
			throw new ExpectedError("Invalid wallet id", Reason.INVALID_ARGUMENT);
		}

		const wallet = await WalletRepository.instance.getWallet(walletId);

		if (!wallet.exists) {
			throw new ExpectedError("Wallet does not exist", Reason.INVALID_ARGUMENT);
		}

		const payoutAmount = moneyToInt(request.payout.amount);

		if (moneyToInt(request.payout.amount) > Math.abs(wallet.get("balance"))) {
			throw new ExpectedError(
				"Amount must be lesser than or equal to balance",
				Reason.INVALID_ARGUMENT
			);
		}

		const payoutId = nanoid();

		// Fuck razorpay for lack of typing
		// TODO: change to payout api
		const data = razorpay.orders.create({
			currency: request.payout.amount?.currencyCode ?? "INR",
			amount: payoutAmount,
		});

		if (!data) {
			throw new ExpectedError("Internal error", Reason.INTERNAL);
		}

		const checkoutInfo = {
			payment_gateway: "razorpay",
			rzp_order_id: data.id,
		};

		const payout = Payout.create({
			name: `${request.parent}/recharges/${payoutId}`,
			amount: request.payout.amount,
			status: Payout_Status.PENDING,
		});

		await PayoutRepository.instance.savePayout(walletId, payout, checkoutInfo);

		return {
			payout,
			checkoutInfo,
		};
	} catch (err) {
		if (err instanceof ExpectedError) {
			throw err;
		}
		throw new ExpectedError("Failed to create payout", Reason.INTERNAL);
	}
}

export default createPayout;

File: api/service/create_payout.go
package service

import (
	"context"
	"errors"
	"math"
	"strings"

	"connectrpc.com/connect"
	"github.com/aidarkhanov/nanoid"
	pb "github.com/ride-app/wallet-service/api/gen/ride/wallet/v1alpha1"
	walletrepository "github.com/ride-app/wallet-service/repositories/wallet"
)

func (service *WalletServiceServer) CreatePayout(ctx context.Context, req *connect.Request[pb.CreatePayoutRequest]) (*connect.Response[pb.CreatePayoutResponse], error) {
	log := service.logger.WithField("method", "CreatePayout")

	if err := req.Msg.Validate(); err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}

	uid := strings.Split(req.Msg.Parent, "/")[1]

	wallet, err := service.walletRepository.GetWallet(ctx, log, uid)

	if err != nil || wallet == nil {
		return nil, connect.NewError(connect.CodeFailedPrecondition, errors.New("parent Wallet not found"))
	}

	if wallet.Balance <= 0 || req.Msg.Payout.Amount > int32(math.Abs(float64(wallet.Balance))) {
		return nil, connect.NewError(connect.CodeInvalidArgument, errors.New("amount must be smaller than wallet balance"))
	}

	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	payoutAccount, err := service.payoutRepository.GetPayoutAccount(ctx, log, uid)

	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	if payoutAccount == nil {
		return nil, connect.NewError(connect.CodeFailedPrecondition, errors.New("payout account not found"))
	}

	payout, err := service.payoutRepository.CreatePayout(ctx, log, payoutAccount, req.Msg.Payout)

	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	transactions := &walletrepository.Transactions{
		uid: &pb.Transaction{
			Type:   pb.Transaction_TYPE_DEBIT,
			Amount: req.Msg.Payout.Amount,
		},
	}

	err = service.walletRepository.CreateTransactions(ctx, log, transactions, nanoid.New())

	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	response := connect.NewResponse(&pb.CreatePayoutResponse{
		Payout: payout,
	})

	if err = response.Msg.Validate(); err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	return response, nil
}
